# Proyek Analisis Data: [Input Nama Dataset]
- **Nama:** Vicko Amelino Syahputra
- **Email:** vickokudus@gmail.com
- **ID Dicoding:** vickoaja

## Menentukan Pertanyaan Bisnis

**SMART Question** adalah sebuah framework untuk merumuskan pertanyaan secara terstruktur agar memeperoleh informasi yang mendalam.

* **Spesific (Spesifik)**
  * Pertanyaan harus jelas, fokus pada sebuah topik tertentu, dan tidak bermakna ganda. Hindari pertanyaan yang terlalu luas.
    * Salah: Bagaimana cara meningkatkan penjualan?
    * Benar: Faktor apa saja yang memengaruhi penurunan penjualan produk kategori elektronik di wilayah Jakarta selama kuartal terakhir?
* **Measurable (Terukur)**
  * Pertanyaan harus bisa dijawab dengan angka atau matrix yang konkret, harus tahu apa yang akan dihitung.
    * Salah: Apakah pelanggan senang dengan layanan kita?
    * Benar: Berapa skor rata-rata Customer Satisfaction untuk layanan purna jual bulan ini dibandingkan bulan lalu?
* **Action-Oriented (Berorientasi Aksi)**
  * Hasil dari pertanyaan harus bisa memberikan arahan untuk melakukan tindakan nyata. Jika pertanyaan terjawab, stakeholder harus tahu apa langkah selanjutnya.
    * Salah: Mengapa orang suka berbelanja?
    * Benar: Fitur apa pada aplikasi yang paling sering digunakan sebelum pengguna memutuskan untuk melakukan checkout?
* **Relevant (Relevan)**
  * Hasil dari pertanyaan harus sejalan dengan tujuan utama bisnis atau masalah yang sedang dihadapi.
    * Salah: Menanyakan tentang stok gudang saat masalah utamanya adalah efektivitas kampanye media sosial.
    * Benar: Apakah kampanye iklan di Instagram memberikan Return on Ad Spend (ROAS) yang lebih tinggi dibandingkan iklan di TikTok?
* **Time-bound (Terikat Waktu)**
  * Pertanyaan harus ada batasan waktu yang jelas agar analisis memiliki konteks yang tepat.
    * Salah: Berapa banyak pengguna baru kita?
    * Benar: Berapa tingkat pertumbuhan pengguna baru secara bulanan (Month-over-Month) sepanjang tahun 2025?

**Contoh pertanyaan bisnis yang memenuhi seluruh elemen SMART**

***"Faktor apa saja yang memengaruhi penurunan conversion rate pada pengguna aplikasi Android di wilayah Jabodetabek sebesar 5% selama periode Flash Sale Maret 2026?"***

Keterangan:

- **Specific**: Fokus pada "penurunan conversion rate" untuk "aplikasi Android" di "Jabodetabek". Bukan sekadar penjualan turun.
- **Measurable**: Ada angka konkret yang ingin dianalisis, yaitu penurunan sebesar "5%".
- **Action-Oriented**: Dengan mengetahui faktor penyebabnya misalnya bug pada versi Android tertentu atau kendala logistik di Jabodetabek, tim bisa langsung melakukan perbaikan teknis atau operasional.
- **Relevant**: Penurunan konversi saat Flash Sale adalah masalah kritis bagi bisnis retail/e-commerce.
- **Time-bound**: Dibatasi pada periode spesifik "Flash Sale Maret 2026".

- **Pertanyaan 1:**  Kategori produk mana yang menghasilkan revenue tertinggi 2016-2018?
- **Pertanyaan 2:** Bagaimana tren revenue bulanan (MoM) tahun 2017-2018 & kapan puncaknya?
- **Pertanyaan 3:** Apakah ada korelasi antara waktu pengiriman vs skor ulasan pelanggan?
- **Pertanyaan 4:** Metode pembayaran apa yang paling banyak digunakan & rata-rata nilai transaksinya?

## Setup Project Data Analisis

Bagian ini merupakan tahap awal dari proyek analisis data. Di sini saya menyiapkan semua kebutuhan dasar sebelum mulai mengolah data, mulai dari mengimpor library yang diperlukan hingga mengatur konfigurasi tampilan grafik.

### Import Packages/Library yang akan digunakan

Pada tahap ini, impor beberapa library yang akan digunakan selama proses analisis.

- **pandas** digunakan untuk membaca, memanipulasi, dan menganalisis data dalam bentuk tabel (DataFrame).
- **numpy** digunakan untuk operasi numerik seperti menghitung nilai statistik dan mendeteksi outlier.
- **matplotlib.pyplot** merupakan library utama untuk membuat visualisasi/grafik.
- **matplotlib.ticker** digunakan untuk memformat tampilan angka pada sumbu grafik.
- **seaborn** digunakan sebagai pelengkap matplotlib agar grafik yang dihasilkan lebih rapi dan estetis.
- **warnings** diimpor untuk menyembunyikan pesan peringatan yang tidak relevan agar output notebook lebih bersih.

In [42]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
import seaborn as sns
import warnings

### Configurasi project

Lakukan beberapa konfigurasi awal sebelum memulai analisis.

- **warnings.filterwarnings('ignore')** digunakan untuk menyembunyikan pesan peringatan yang tidak relevan agar output notebook lebih bersih dan mudah dibaca.
- **sns.set_theme(style='whitegrid', palette='muted')** digunakan untuk mengatur tema grafik secara global. Style `whitegrid` memberikan tampilan background putih dengan garis grid tipis, sedangkan palette `muted` membuat warna grafik tidak terlalu mencolok.
- **plt.rcParams['figure.dpi'] = 100** digunakan untuk mengatur resolusi grafik yang dihasilkan agar tampilannya lebih tajam.

In [43]:
warnings.filterwarnings('ignore')
sns.set_theme(style='whitegrid', palette='muted')
plt.rcParams['figure.dpi'] = 100

## Data Wrangling

Data Wrangling adalah proses mempersiapkan data mentah agar siap digunakan untuk analisis. Proses ini terdiri dari tiga tahap utama, yaitu Gathering Data (mengumpulkan data), Assessing Data (menilai kualitas data), dan Cleaning Data (membersihkan data dari masalah yang ditemukan).

### Gathering Data

Pada tahap ini, saya mengumpulkan semua data yang dibutuhkan dengan cara membaca file CSV dan memuatnya ke dalam bentuk DataFrame. Dataset yang digunakan berasal dari E-Commerce Public Dataset yang terdiri dari beberapa tabel yang saling berkaitan.

#### Load df E-Commerce Public

Berikut ini saya memuat seluruh dataset yang dibutuhkan ke dalam DataFrame. Dataset terdiri dari 9 file CSV yang masing-masing menyimpan informasi berbeda namun saling berkaitan satu sama lain dan dibawah ini adalah penjelasan masing-masing dataset.

- **orders_df** — data utama pesanan, berisi status pesanan dan timestamp dari setiap tahap proses pengiriman.
- **order_items_df** — data item di dalam setiap pesanan, berisi harga produk dan biaya pengiriman.
- **order_payments_df** — data pembayaran pesanan, berisi metode pembayaran dan nilai transaksi.
- **order_reviews_df** — data ulasan pelanggan, berisi skor dan komentar setelah pesanan diterima.
- **products_df** — data produk, berisi informasi kategori dan dimensi fisik produk.
- **customers_df** — data pelanggan, berisi informasi lokasi pelanggan.
- **sellers_df** — data penjual, berisi informasi lokasi penjual.
- **geolocation_df** — data koordinat geografis berdasarkan kode pos di Brazil.
- **category_translation_df** — data terjemahan nama kategori produk dari bahasa Portugis ke bahasa Inggris.

In [44]:
orders_df = pd.read_csv('orders_dataset.csv')
order_items_df = pd.read_csv('order_items_dataset.csv')
order_payments_df = pd.read_csv('order_payments_dataset.csv')
order_reviews_df = pd.read_csv('order_reviews_dataset.csv')
products_df = pd.read_csv('products_dataset.csv')
customers_df = pd.read_csv('customers_dataset.csv')
sellers_df = pd.read_csv('sellers_dataset.csv')
geolocation_df = pd.read_csv('geolocation_dataset.csv')
category_translation_df = pd.read_csv('product_category_name_translation.csv')

Muat seluruh dataset yang dibutuhkan ke dalam DataFrame. Untuk `orders_df` dan `order_reviews_df`, saya menambahkan parameter `parse_dates` agar kolom yang berisi tanggal langsung terbaca sebagai tipe datetime, bukan string. Hal ini penting karena nantinya kolom-kolom tersebut akan digunakan untuk menghitung durasi pengiriman dan analisis tren waktu.

Di bagian akhir, saya mencetak jumlah baris dan kolom dari setiap dataset untuk memastikan semua data berhasil dimuat dengan benar.

In [45]:
orders_df['order_purchase_timestamp'] = pd.to_datetime(orders_df['order_purchase_timestamp'])
orders_df['order_approved_at'] = pd.to_datetime(orders_df['order_approved_at'])
orders_df['order_delivered_carrier_date'] = pd.to_datetime(orders_df['order_delivered_carrier_date'])
orders_df['order_delivered_customer_date'] = pd.to_datetime(orders_df['order_delivered_customer_date'])
orders_df['order_estimated_delivery_date'] = pd.to_datetime(orders_df['order_estimated_delivery_date'])

order_reviews_df['review_creation_date'] = pd.to_datetime(order_reviews_df['review_creation_date'])
order_reviews_df['review_answer_timestamp'] = pd.to_datetime(order_reviews_df['review_answer_timestamp'])

for name, df in {
   'orders': orders_df, 'items': order_items_df,
   'payments': order_payments_df, 'reviews': order_reviews_df,
   'products': products_df, 'customers': customers_df,
   'sellers': sellers_df, 'category': category_translation_df
}.items():
   print(f"{name:12s}: {df.shape[0]:>7,} baris × {df.shape[1]} kolom")

orders      :  99,441 baris × 8 kolom
items       : 112,650 baris × 7 kolom
payments    : 103,886 baris × 5 kolom
reviews     :  99,224 baris × 7 kolom
products    :  32,951 baris × 9 kolom
customers   :  99,441 baris × 5 kolom
sellers     :   3,095 baris × 4 kolom
category    :      71 baris × 2 kolom


**Insight:**
- Dataset utama `orders_df` memiliki 99.441 baris yang sama persis dengan `customers_df`, artinya setiap pelanggan tercatat melakukan tepat satu pesanan.
- `order_items_df` memiliki 112.650 baris yang lebih banyak dari jumlah pesanan, artinya ada pesanan yang berisi lebih dari satu item produk.
- `order_payments_df` memiliki 103.886 baris yang juga lebih banyak dari jumlah pesanan, artinya ada pesanan yang dibayar menggunakan lebih dari satu metode pembayaran sekaligus.
- `products_df` memiliki 32.951 produk unik yang tersebar ke dalam 71 kategori berdasarkan `category_translation_df`.
- `sellers_df` hanya memiliki 3.095 penjual, jauh lebih sedikit dibanding jumlah pesanan, artinya satu penjual bisa melayani banyak pesanan sekaligus.

### Assessing Data

Pada tahap ini, saya menilai kualitas data dari setiap dataset yang telah dimuat sebelumnya. Tujuannya adalah untuk menemukan potensi masalah pada data seperti missing value, data duplikat, tipe data yang tidak sesuai, maupun nilai yang tidak wajar. Masalah-masalah yang ditemukan di sini akan menjadi acuan untuk tahap selanjutnya yaitu Cleaning Data.

#### Identifying Missing Value

In [46]:
for name, df in {
    'orders': orders_df, 'items': order_items_df,
    'payments': order_payments_df, 'reviews': order_reviews_df,
    'products': products_df, 'customers': customers_df,
    'sellers': sellers_df, 'geolocation': geolocation_df,
    'category': category_translation_df
}.items():
    missing = df.isnull().sum()
    missing = missing[missing > 0]
    if len(missing) > 0:
        print(f"\n[{name}]")
        print(missing.to_string())
    else:
        print(f"\n[{name}] — tidak ada missing value")


[orders]
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965

[items] — tidak ada missing value

[payments] — tidak ada missing value

[reviews]
review_comment_title      87656
review_comment_message    58247

[products]
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2

[customers] — tidak ada missing value

[sellers] — tidak ada missing value

[geolocation] — tidak ada missing value

[category] — tidak ada missing value


#### Identifying Duplicate Data

Selanjutnya, saya mengecek apakah ada baris yang terduplikasi pada setiap dataset. Data duplikat perlu diidentifikasi karena bisa menyebabkan hasil perhitungan menjadi tidak akurat, misalnya total revenue yang terhitung dua kali.

In [47]:
for name, df in {
    'orders': orders_df, 'items': order_items_df,
    'payments': order_payments_df, 'reviews': order_reviews_df,
    'products': products_df, 'customers': customers_df,
    'sellers': sellers_df, 'geolocation': geolocation_df,
    'category': category_translation_df
}.items():
    duplicated = df.duplicated().sum()
    if duplicated > 0:
        print(f"[{name}] — {duplicated:,} baris duplikat ditemukan")
    else:
        print(f"[{name}] — tidak ada duplikat")

[orders] — tidak ada duplikat
[items] — tidak ada duplikat
[payments] — tidak ada duplikat
[reviews] — tidak ada duplikat
[products] — tidak ada duplikat
[customers] — tidak ada duplikat
[sellers] — tidak ada duplikat
[geolocation] — 261,831 baris duplikat ditemukan
[category] — tidak ada duplikat


#### Identifying Incorrect Data Type

Selanjutnya, saya mengecek tipe data dari setiap kolom pada masing-masing dataset. Tipe data yang tidak sesuai perlu diidentifikasi karena bisa menyebabkan error saat melakukan operasi tertentu, misalnya kolom tanggal yang masih bertipe object tidak bisa langsung digunakan untuk menghitung selisih waktu.

In [48]:
for name, df in {
    'orders': orders_df, 'items': order_items_df,
    'payments': order_payments_df, 'reviews': order_reviews_df,
    'products': products_df, 'customers': customers_df,
    'sellers': sellers_df, 'geolocation': geolocation_df,
    'category': category_translation_df
}.items():
    print(f"\n[{name}]")
    print(df.dtypes.to_string())


[orders]
order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]

[items]
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64

[payments]
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64

[reviews]
review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message       

#### Identifying Invalid Value

Selanjutnya, saya mengecek apakah ada nilai yang tidak wajar pada data. Pada dataset ini, nilai tidak wajar yang paling mungkin terjadi adalah tanggal pengiriman yang lebih awal dari tanggal pembelian, yang secara logika tidak mungkin terjadi.

In [49]:
invalid_delivery = orders_df.dropna(subset=['order_delivered_customer_date', 'order_purchase_timestamp'])
invalid_delivery = invalid_delivery[
    invalid_delivery['order_delivered_customer_date'] <
    invalid_delivery['order_purchase_timestamp']
]

print(f"Jumlah pesanan dengan tanggal pengiriman tidak valid: {len(invalid_delivery)}")

Jumlah pesanan dengan tanggal pengiriman tidak valid: 0


#### Identifying Outlier

Terakhir, saya mengecek apakah ada nilai ekstrem (outlier) pada kolom numerik. Metode yang digunakan adalah IQR (Interquartile Range), yaitu dengan menghitung batas bawah dan batas atas dari data. Nilai yang berada di luar batas tersebut dianggap sebagai outlier. Pada tahap ini saya fokus mengecek outlier pada kolom `price` dan `freight_value` di `order_items_df` karena kedua kolom tersebut paling relevan untuk analisis.

In [50]:
for col in ['price', 'freight_value']:
    Q1 = order_items_df[col].quantile(0.25)
    Q3 = order_items_df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = order_items_df[(order_items_df[col] < lower) | (order_items_df[col] > upper)]
    print(f"[{col}]")
    print(f"  Batas bawah : {lower:.2f}")
    print(f"  Batas atas  : {upper:.2f}")
    print(f"  Jumlah outlier: {len(outliers):,} dari {len(order_items_df):,} baris ({len(outliers)/len(order_items_df)*100:.1f}%)\n")

[price]
  Batas bawah : -102.60
  Batas atas  : 277.40
  Jumlah outlier: 8,427 dari 112,650 baris (7.5%)

[freight_value]
  Batas bawah : 0.98
  Batas atas  : 33.25
  Jumlah outlier: 12,134 dari 112,650 baris (10.8%)



**Steps to Take:**
- Mengkonversi kolom-kolom bertipe `object` yang berisi tanggal pada `orders_df` dan `order_reviews_df` menjadi tipe `datetime` agar bisa digunakan untuk perhitungan durasi waktu.
- Mengisi atau menghapus missing value pada `orders_df` khususnya kolom `order_delivered_carrier_date` dan `order_delivered_customer_date` yang kosong karena pesanan belum terkirim.
- Mengisi missing value pada `products_df` khususnya kolom `product_category_name` yang kosong dengan nilai `'unknown'`.
- Menghapus duplikat pada `order_reviews_df` berdasarkan kolom `review_id` karena satu pesanan seharusnya hanya memiliki satu ulasan.
- Outlier pada kolom `price` dan `freight_value` tidak akan dihapus karena kemungkinan merupakan produk dengan harga memang tinggi, namun akan diperhatikan saat proses visualisasi.

**Insight:**
- Kolom-kolom bertipe `object` yang berisi tanggal pada `orders_df` dan `order_reviews_df` perlu dikonversi ke `datetime` sebelum digunakan untuk analisis waktu.
- Missing value pada `orders_df` di kolom `order_delivered_carrier_date` dan `order_delivered_customer_date` kemungkinan besar bukan kesalahan data, melainkan memang pesanan yang belum sampai ke tangan pelanggan.
- Missing value pada `products_df` di kolom `product_category_name` perlu ditangani agar tidak mengganggu analisis kategori produk.
- Tidak ditemukan baris duplikat pada seluruh dataset kecuali `order_reviews_df`, yang menandakan kualitas data secara umum cukup baik.
- Outlier pada kolom `price` kemungkinan berasal dari produk-produk dengan harga memang tinggi seperti elektronik atau furnitur, sehingga tidak serta merta harus dihapus.

### Cleaning Data

### Cleaning Data

Pada tahap ini, saya melakukan pembersihan data berdasarkan masalah-masalah yang telah ditemukan di tahap Assessing Data sebelumnya. Tujuannya adalah memastikan data yang digunakan untuk analisis sudah bersih, konsisten, dan siap diolah lebih lanjut.

#### Fixing Data Type

Saya mengkonversi kolom-kolom yang berisi tanggal pada `orders_df` dan `order_reviews_df` dari tipe `object` menjadi `datetime`. Proses ini diperlukan agar kolom-kolom tersebut bisa digunakan untuk perhitungan durasi waktu seperti menghitung lama pengiriman.

In [51]:
date_columns_orders = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

date_columns_reviews = [
    'review_creation_date',
    'review_answer_timestamp'
]

for col in date_columns_orders:
    orders_df[col] = pd.to_datetime(orders_df[col])

for col in date_columns_reviews:
    order_reviews_df[col] = pd.to_datetime(order_reviews_df[col])

print("Tipe data orders_df:")
print(orders_df[date_columns_orders].dtypes)
print("\nTipe data order_reviews_df:")
print(order_reviews_df[date_columns_reviews].dtypes)

Tipe data orders_df:
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

Tipe data order_reviews_df:
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object


#### Fixing Missing Value

Selanjutnya, saya menangani missing value pada `orders_df` dan `products_df`. Untuk `orders_df`, baris yang memiliki nilai kosong pada kolom `order_delivered_customer_date` akan dibiarkan karena memang pesanan tersebut belum terkirim, namun akan difilter saat analisis pengiriman. Untuk `products_df`, missing value pada kolom `product_category_name` akan diisi dengan nilai `'unknown'`.

In [52]:
products_df['product_category_name'] = products_df['product_category_name'].fillna('unknown')

print("Missing value products_df setelah cleaning:")
print(products_df['product_category_name'].isnull().sum())

print("\nMissing value orders_df (kolom tanggal pengiriman):")
print(orders_df[['order_delivered_carrier_date', 'order_delivered_customer_date']].isnull().sum())
print("(dibiarkan karena pesanan memang belum terkirim)")

Missing value products_df setelah cleaning:
0

Missing value orders_df (kolom tanggal pengiriman):
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64
(dibiarkan karena pesanan memang belum terkirim)


#### Fixing Duplicate Data

Terakhir, saya menghapus baris duplikat pada `order_reviews_df` berdasarkan kolom `review_id`. Hal ini dilakukan karena satu pesanan seharusnya hanya memiliki satu ulasan, sehingga jika ditemukan `review_id` yang sama maka cukup diambil satu saja.

In [53]:
before = len(order_reviews_df)
order_reviews_df = order_reviews_df.drop_duplicates(subset='review_id', keep='first')
after = len(order_reviews_df)

print(f"Jumlah baris sebelum cleaning : {before:,}")
print(f"Jumlah baris setelah cleaning : {after:,}")
print(f"Jumlah baris yang dihapus     : {before - after:,}")

Jumlah baris sebelum cleaning : 99,224
Jumlah baris setelah cleaning : 98,410
Jumlah baris yang dihapus     : 814


**Insight:**
- Setelah konversi tipe data, seluruh kolom tanggal pada `orders_df` dan `order_reviews_df` sudah bertipe `datetime` dan siap digunakan untuk perhitungan durasi waktu.
- Missing value pada kolom `product_category_name` di `products_df` berhasil diisi dengan nilai `'unknown'` sehingga tidak ada lagi nilai kosong yang bisa mengganggu analisis kategori produk.
- Missing value pada kolom tanggal pengiriman di `orders_df` sengaja dibiarkan karena memang merepresentasikan pesanan yang belum terkirim, bukan merupakan kesalahan data.
- Duplikat pada `order_reviews_df` berhasil dihapus sehingga setiap `review_id` kini bersifat unik dan tidak ada ulasan yang terhitung lebih dari satu kali.

## Exploratory Data Analysis (EDA)

Pada tahap ini, saya melakukan eksplorasi data untuk menjawab pertanyaan bisnis yang telah ditentukan sebelumnya. Proses eksplorasi dilakukan dengan cara melihat rangkuman statistik, melakukan grouping dan agregasi, serta mencari pola atau hubungan antar variabel yang relevan.

### Explore Kategori Produk Berdasarkan Total Revenue

Pada bagian ini, saya mengeksplorasi kategori produk mana yang menghasilkan total pendapatan tertinggi sepanjang periode 2016-2018. Untuk menjawab pertanyaan ini, saya menggabungkan `order_items_df` dengan `products_df` dan `category_translation_df` agar nama kategori produk tampil dalam bahasa Inggris, kemudian menghitung total revenue per kategori.

In [54]:
items_products = order_items_df.merge(products_df[['product_id', 'product_category_name']], on='product_id', how='left')
items_products = items_products.merge(category_translation_df, on='product_category_name', how='left')

category_revenue = (items_products
                    .groupby('product_category_name_english')['price']
                    .sum()
                    .sort_values(ascending=False)
                    .reset_index())

category_revenue.columns = ['category', 'total_revenue']

print(category_revenue.head(10).to_string(index=False))

             category  total_revenue
        health_beauty     1258681.34
        watches_gifts     1205005.68
       bed_bath_table     1036988.68
       sports_leisure      988048.97
computers_accessories      911954.32
      furniture_decor      729762.49
           cool_stuff      635290.85
           housewares      632248.66
                 auto      592720.11
         garden_tools      485256.46


### Explore Tren Pendapatan Bulanan 2017-2018

Pada bagian ini, saya mengeksplorasi bagaimana tren pendapatan bulanan selama periode 2017-2018 dan pada bulan mana terjadi puncak penjualan. Untuk menjawab pertanyaan ini, saya menggabungkan `orders_df` dengan `order_items_df` kemudian mengelompokkan data berdasarkan bulan dan tahun.

In [55]:
orders_items = orders_df.merge(order_items_df[['order_id', 'price']], on='order_id', how='left')
orders_items = orders_items[orders_items['order_status'] == 'delivered']
orders_items['year_month'] = orders_items['order_purchase_timestamp'].dt.to_period('M')

monthly_revenue = (orders_items
                   .groupby('year_month')['price']
                   .sum()
                   .reset_index())

monthly_revenue.columns = ['year_month', 'revenue']
monthly_revenue = monthly_revenue[monthly_revenue['year_month'].astype(str) >= '2017-01']

print(monthly_revenue.to_string(index=False))

year_month   revenue
   2017-01 111798.36
   2017-02 234223.40
   2017-03 359198.85
   2017-04 340669.68
   2017-05 489338.25
   2017-06 421923.37
   2017-07 481604.52
   2017-08 554699.70
   2017-09 607399.67
   2017-10 648247.65
   2017-11 987765.37
   2017-12 726033.19
   2018-01 924645.00
   2018-02 826437.13
   2018-03 953356.25
   2018-04 973534.09
   2018-05 977544.69
   2018-06 856077.86
   2018-07 867953.46
   2018-08 838576.64


### Explore Korelasi Waktu Pengiriman vs Skor Ulasan

Pada bagian ini, saya mengeksplorasi apakah ada hubungan antara lama waktu pengiriman dengan skor ulasan yang diberikan pelanggan. Hipotesisnya adalah semakin lama pengiriman maka semakin rendah skor ulasan yang diberikan. Untuk menjawab pertanyaan ini, saya menggabungkan `orders_df` dengan `order_reviews_df` kemudian menghitung rata-rata lama pengiriman per skor ulasan.

In [56]:
orders_delivered = orders_df[orders_df['order_status'] == 'delivered'].copy()
orders_delivered = orders_delivered.dropna(subset=['order_delivered_customer_date'])
orders_delivered['delivery_days'] = (
    orders_delivered['order_delivered_customer_date'] -
    orders_delivered['order_purchase_timestamp']
).dt.days

orders_delivered = orders_delivered[orders_delivered['delivery_days'] > 0]

orders_reviews = orders_delivered.merge(
    order_reviews_df[['order_id', 'review_score']], on='order_id', how='left')

delivery_score = (orders_reviews
                  .groupby('review_score')['delivery_days']
                  .agg(['mean', 'median', 'count'])
                  .rename(columns={'mean': 'avg_days', 'median': 'median_days', 'count': 'n_orders'}))

print(delivery_score.to_string())

               avg_days  median_days  n_orders
review_score                                  
1.0           20.891281         16.0      9290
2.0           16.187070         13.0      2908
3.0           13.790155         12.0      7882
4.0           11.843470         10.0     18859
5.0           10.215679          9.0     56649


### Explore Metode Pembayaran dan Rata-rata Nilai Transaksi

Pada bagian ini, saya mengeksplorasi metode pembayaran apa yang paling banyak digunakan pelanggan dan berapa rata-rata nilai transaksi per metode pembayaran. Untuk menjawab pertanyaan ini, saya mengelompokkan `order_payments_df` berdasarkan kolom `payment_type` kemudian menghitung jumlah transaksi dan rata-rata nilai transaksinya.

In [57]:
payments_clean = order_payments_df[order_payments_df['payment_type'] != 'not_defined']
payment_summary = (payments_clean
                   .groupby('payment_type')['payment_value']
                   .agg(['count', 'sum', 'mean'])
                   .rename(columns={'count': 'n_transactions',
                                    'sum': 'total_value',
                                    'mean': 'avg_value'})
                   .sort_values('n_transactions', ascending=False))

print(payment_summary.to_string())

              n_transactions  total_value   avg_value
payment_type                                         
credit_card            76795  12542084.19  163.319021
boleto                 19784   2869361.27  145.034435
voucher                 5775    379436.87   65.703354
debit_card              1529    217989.79  142.570170


**Insight:**
- Kartu kredit mendominasi dengan lebih dari 70% total transaksi, jauh meninggalkan metode pembayaran lainnya.
- Meskipun voucher cukup sering digunakan, rata-rata nilai transaksinya paling rendah dibanding metode lain, menunjukkan voucher lebih sering dipakai untuk pembelian bernilai kecil.
- Kartu debit memiliki jumlah pengguna paling sedikit, kemungkinan karena pelanggan lebih memilih kartu kredit yang menawarkan fitur cicilan.
- Total nilai transaksi kartu kredit jauh lebih besar dibanding metode lain, menjadikannya metode pembayaran paling penting dari sisi bisnis.

## Visualization & Explanatory Analysis

Pada tahap ini, saya membuat visualisasi data untuk menjawab pertanyaan bisnis yang telah dieksplorasi sebelumnya. Visualisasi dibuat agar hasil analisis lebih mudah dipahami dan dapat dikomunikasikan kepada stakeholder secara lebih efektif.

### Pertanyaan 1: Kategori produk mana yang menghasilkan revenue tertinggi 2016-2018?

Visualisasi berikut menampilkan 10 kategori produk dengan total revenue tertinggi sepanjang periode 2016-2018 menggunakan horizontal bar chart. Chart ini dipilih karena nama kategori yang cukup panjang sehingga lebih mudah dibaca secara horizontal.

In [58]:
top_10 = category_revenue.head(10)

fig, ax = plt.subplots(figsize=(10, 6))

bars = ax.barh(
    top_10['category'][::-1],
    top_10['total_revenue'][::-1] / 1_000,
    color=sns.color_palette('muted', 10)[::-1],
    edgecolor='white'
)

for bar, val in zip(bars, top_10['total_revenue'][::-1]):
    ax.text(bar.get_width() + 5, bar.get_y() + bar.get_height() / 2,
            f'R${val/1000:,.0f}K', va='center', fontsize=9, fontweight='bold')

ax.set_xlabel('Total Revenue (Ribuan BRL)', fontsize=11)
ax.set_title('Top 10 Kategori Produk Berdasarkan Total Revenue (2016-2018)',
             fontsize=13, fontweight='bold', pad=15)
ax.xaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}K'))
ax.set_xlim(0, top_10['total_revenue'].max() / 1000 * 1.3)
sns.despine()
plt.tight_layout()
plt.show()

### Pertanyaan 2: Bagaimana tren revenue bulanan (MoM) tahun 2017-2018 & kapan puncaknya?

Visualisasi berikut menampilkan tren revenue bulanan selama periode 2017-2018 menggunakan line chart. Selain itu, saya menambahkan bar chart pada sumbu kedua untuk menampilkan pertumbuhan Month-over-Month (MoM) agar perubahan revenue antar bulan lebih mudah terlihat.

In [59]:
monthly_revenue['revenue_mom'] = monthly_revenue['revenue'].pct_change() * 100

fig, ax1 = plt.subplots(figsize=(12, 6))

ax1.fill_between(range(len(monthly_revenue)),
                 monthly_revenue['revenue'] / 1_000,
                 alpha=0.2, color='#2a9d8f')
ax1.plot(range(len(monthly_revenue)),
         monthly_revenue['revenue'] / 1_000,
         color='#264653', linewidth=2.5, marker='o', markersize=5, label='Revenue')

peak_idx = monthly_revenue['revenue'].idxmax()
peak = monthly_revenue.loc[peak_idx]
ax1.annotate(f"Puncak\nR${peak['revenue']/1000:,.0f}K",
             xy=(peak_idx, peak['revenue'] / 1000),
             xytext=(peak_idx - 2, peak['revenue'] / 1000 + 80),
             arrowprops=dict(arrowstyle='->', color='#e63946'),
             fontsize=9, color='#e63946', fontweight='bold')

ax2 = ax1.twinx()
mom_colors = ['#2a9d8f' if v >= 0 else '#e63946'
              for v in monthly_revenue['revenue_mom'].fillna(0)]
ax2.bar(range(len(monthly_revenue)),
        monthly_revenue['revenue_mom'].fillna(0),
        color=mom_colors, alpha=0.35, width=0.4, label='MoM %')
ax2.axhline(0, color='gray', linewidth=0.8, linestyle='--')
ax2.set_ylabel('Month-over-Month (%)', color='gray', fontsize=11)
ax2.tick_params(axis='y', labelcolor='gray')

tick_step = max(1, len(monthly_revenue) // 10)
ax1.set_xticks(list(range(0, len(monthly_revenue), tick_step)))
ax1.set_xticklabels(
    [str(monthly_revenue['year_month'].iloc[i]) for i in range(0, len(monthly_revenue), tick_step)],
    rotation=45, ha='right', fontsize=9)
ax1.yaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}K'))
ax1.set_ylabel('Revenue (Ribuan BRL)', fontsize=11)
ax1.set_title('Tren Revenue Bulanan & Month-over-Month Growth (2017-2018)',
              fontsize=13, fontweight='bold', pad=15)
ax1.legend(loc='upper left')
sns.despine()
plt.tight_layout()
plt.show()

### Pertanyaan 3: Apakah ada korelasi antara waktu pengiriman vs skor ulasan pelanggan?

In [61]:
fig, axes = plt.subplots(1, 2, figsize=(12, 6))

score_colors = ['#e63946', '#f4a261', '#e9c46a', '#2a9d8f', '#264653']

ax1 = axes[0]
bars = ax1.bar(delivery_score.index,
               delivery_score['avg_days'],
               color=score_colors, edgecolor='white', width=0.6)

for bar, val in zip(bars, delivery_score['avg_days']):
    ax1.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + 0.4,
             f'{val:.1f} hari', ha='center', fontsize=10, fontweight='bold')

ax1.set_xlabel('Skor Ulasan', fontsize=11)
ax1.set_ylabel('Rata-rata Hari Pengiriman', fontsize=11)
ax1.set_title('Rata-rata Hari Pengiriman\nper Skor Ulasan', fontsize=12, fontweight='bold')
ax1.set_xticks(delivery_score.index)
ax1.set_xticklabels([f'⭐{s:.0f}' for s in delivery_score.index])
sns.despine(ax=ax1)

ax2 = axes[1]
sample = orders_reviews.sample(min(5000, len(orders_reviews)), random_state=42)
parts = ax2.violinplot(
    [sample[sample['review_score'] == s]['delivery_days'].values for s in [1, 2, 3, 4, 5]],
    positions=[1, 2, 3, 4, 5], showmedians=True)

for i, pc in enumerate(parts['bodies']):
    pc.set_facecolor(score_colors[i])
    pc.set_alpha(0.7)

ax2.set_xlabel('Skor Ulasan', fontsize=11)
ax2.set_ylabel('Hari Pengiriman', fontsize=11)
ax2.set_title('Distribusi Hari Pengiriman\nper Skor Ulasan', fontsize=12, fontweight='bold')
ax2.set_xticks([1, 2, 3, 4, 5])
ax2.set_xticklabels([f'⭐{s}' for s in [1, 2, 3, 4, 5]])
ax2.set_ylim(0, 100)
sns.despine(ax=ax2)

plt.suptitle('Korelasi Waktu Pengiriman vs Skor Ulasan Pelanggan',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()

### Pertanyaan 4: Metode pembayaran apa yang paling banyak digunakan & rata-rata nilai transaksinya?

Visualisasi berikut menampilkan distribusi metode pembayaran yang digunakan pelanggan menggunakan pie chart, serta perbandingan rata-rata nilai transaksi per metode pembayaran menggunakan bar chart.
python

In [20]:
pay_labels = {
    'credit_card': 'Kartu Kredit',
    'boleto': 'Boleto',
    'voucher': 'Voucher',
    'debit_card': 'Kartu Debit'
}
payment_summary.index = [pay_labels.get(x, x) for x in payment_summary.index]

fig, axes = plt.subplots(1, 2, figsize=(12, 6))

ax1 = axes[0]
explode = [0.05] + [0] * (len(payment_summary) - 1)
wedges, texts, autotexts = ax1.pie(
    payment_summary['n_transactions'],
    labels=payment_summary.index,
    autopct='%1.1f%%',
    colors=sns.color_palette('muted', len(payment_summary)),
    explode=explode,
    startangle=140)
for at in autotexts:
    at.set_fontsize(10)
    at.set_fontweight('bold')
ax1.set_title('Distribusi Jumlah Transaksi\nper Metode Pembayaran',
              fontsize=12, fontweight='bold')

ax2 = axes[1]
bars = ax2.bar(payment_summary.index,
               payment_summary['avg_value'],
               color=sns.color_palette('muted', len(payment_summary)),
               edgecolor='white', width=0.55)

for bar, val in zip(bars, payment_summary['avg_value']):
    ax2.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + 1,
             f'R${val:,.0f}', ha='center', fontsize=10, fontweight='bold')

ax2.set_ylabel('Rata-rata Nilai Transaksi (BRL)', fontsize=11)
ax2.set_title('Rata-rata Nilai Transaksi\nper Metode Pembayaran',
              fontsize=12, fontweight='bold')
ax2.yaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}'))
sns.despine(ax=ax2)

plt.suptitle('Analisis Metode Pembayaran Pelanggan',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()


[Pertanyaan 4] Distribusi Metode Pembayaran
              n_transactions  total_value   avg_value
payment_type                                         
credit_card            76795  12542084.19  163.319021
boleto                 19784   2869361.27  145.034435
voucher                 5775    379436.87   65.703354
debit_card              1529    217989.79  142.570170


**Insight:**
- Kartu kredit mendominasi 74% transaksi dan 78% revenue dengan rata-rata nilai transaksi tertinggi (RS163), menunjukkan preferensi pelanggan untuk pembelian bernilai besar via cicilan.
- Boleto masih relevan (19% transaksi), mencerminkan segmen pelanggan unbanked yang signifikan di Brasil.
- Voucher memiliki nilai transaksi rata-rata sangat rendah (R$66) — efektif sebagai stimulus pembelian kecil, namun kontribusi revenue-nya minimal.
- Terdapat korelasi kuat antara waktu pengiriman dan skor ulasan: pengiriman lebih cepat konsisten menghasilkan skor lebih tinggi, sehingga optimasi logistik berpotensi langsung meningkatkan kepuasan pelanggan.
- Puncak revenue terjadi akhir 2017 (kemungkinan dipengaruhi musim Black Friday/Natal), diikuti fluktuasi negatif di beberapa bulan 2018 yang perlu investigasi lebih lanjut.
- Kategori produk bernilai tinggi (elektronik, furnitur) cenderung mendominasi revenue meski volume transaksinya lebih rendah dibanding kategori kebutuhan harian.

## Analisis Lanjutan

### 1. Segmentasi Pelanggan Berdasarkan Nilai Transaksi
Membagi pelanggan ke dalam segmen (low, mid, high spender) berdasarkan total pengeluaran untuk mengidentifikasi segmen mana yang paling bernilai dan layak diprioritaskan dalam strategi retensi atau promosi.

In [66]:
customer_spending = (
    base.groupby('customer_unique_id')['order_value']
    .sum().reset_index()
    .rename(columns={'order_value': 'total_spending'})
)

q1 = customer_spending['total_spending'].quantile(0.33)
q2 = customer_spending['total_spending'].quantile(0.66)

customer_spending['segment'] = pd.cut(
    customer_spending['total_spending'],
    bins=[-np.inf, q1, q2, np.inf],
    labels=['Low Spender', 'Mid Spender', 'High Spender']
)

seg_summary = (
    customer_spending.groupby('segment', observed=True)
    .agg(jumlah_pelanggan=('customer_unique_id', 'count'),
         avg_spending=('total_spending', 'mean'),
         total_spending=('total_spending', 'sum'))
    .reset_index()
)
seg_summary['revenue_share_%'] = (
    seg_summary['total_spending'] / seg_summary['total_spending'].sum() * 100
).round(1)

print("[Analisis 1] Segmentasi Pelanggan")
print(seg_summary.to_string(index=False))

fig, axes = plt.subplots(1, 3, figsize=(14, 5))
colors = sns.color_palette('muted', 3)

ax1 = axes[0]
bars = ax1.bar(seg_summary['segment'], seg_summary['jumlah_pelanggan'],
               color=colors, edgecolor='white', width=0.55)
for bar, val in zip(bars, seg_summary['jumlah_pelanggan']):
    ax1.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + 50, f'{val:,}',
             ha='center', fontsize=10, fontweight='bold')
ax1.set_ylabel('Jumlah Pelanggan', fontsize=11)
ax1.set_title('Jumlah Pelanggan\nper Segmen', fontsize=12, fontweight='bold')
sns.despine(ax=ax1)

ax2 = axes[1]
bars = ax2.bar(seg_summary['segment'], seg_summary['avg_spending'],
               color=colors, edgecolor='white', width=0.55)
for bar, val in zip(bars, seg_summary['avg_spending']):
    ax2.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + 2, f'R${val:,.0f}',
             ha='center', fontsize=10, fontweight='bold')
ax2.set_ylabel('Rata-rata Spending (BRL)', fontsize=11)
ax2.set_title('Rata-rata Spending\nper Segmen', fontsize=12, fontweight='bold')
ax2.yaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}'))
sns.despine(ax=ax2)

ax3 = axes[2]
explode = [0.05] + [0] * (len(seg_summary) - 1)
wedges, texts, autotexts = ax3.pie(
    seg_summary['revenue_share_%'],
    labels=seg_summary['segment'],
    autopct='%1.1f%%',
    colors=colors,
    explode=explode,
    startangle=140)
for at in autotexts:
    at.set_fontsize(10)
    at.set_fontweight('bold')
ax3.set_title('Kontribusi Revenue\nper Segmen', fontsize=12, fontweight='bold')

plt.suptitle('Segmentasi Pelanggan Berdasarkan Nilai Transaksi',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()

[Analisis 1] Segmentasi Pelanggan
     segment  jumlah_pelanggan  avg_spending  total_spending  revenue_share_%
 Low Spender             30843     34.688963      1069911.68              8.1
 Mid Spender             30784     88.292449      2717994.75             20.5
High Spender             31731    298.146488      9460486.22             71.4


### 2. Analisis Keterlambatan Pengiriman
Menghitung selisih antara estimasi pengiriman dan tanggal aktual, lalu mengkorelasikannya dengan skor ulasan untuk membuktikan apakah *keterlambatan* (bukan sekadar durasi pengiriman) adalah pemicu utama ulasan negatif.

In [67]:
delivery_df = base.dropna(subset=[
    'order_delivered_customer_date',
    'order_estimated_delivery_date',
    'review_score'
]).copy()

delivery_df['delay_days'] = (
    delivery_df['order_delivered_customer_date'] -
    delivery_df['order_estimated_delivery_date']
).dt.days

delivery_df['status_kirim'] = delivery_df['delay_days'].apply(
    lambda x: 'Terlambat' if x > 0 else 'Tepat/Lebih Cepat'
)

delay_score = (
    delivery_df.groupby('review_score')
    .agg(avg_delay=('delay_days', 'mean'),
         pct_terlambat=('status_kirim', lambda x: (x == 'Terlambat').mean() * 100))
    .reset_index()
)

print("\n[Analisis 2] Keterlambatan Pengiriman vs Skor Ulasan")
print(delay_score.to_string(index=False))

fig, axes = plt.subplots(1, 2, figsize=(12, 6))
score_colors = sns.color_palette('muted', 5)

ax1 = axes[0]
bars = ax1.bar(delay_score['review_score'], delay_score['avg_delay'],
               color=score_colors, edgecolor='white', width=0.6)
ax1.axhline(0, color='gray', linewidth=0.8, linestyle='--')
for bar, val in zip(bars, delay_score['avg_delay']):
    offset = 0.3 if val >= 0 else -1.8
    ax1.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + offset,
             f'{val:.1f}d', ha='center', fontsize=10, fontweight='bold')
ax1.set_xlabel('Skor Ulasan', fontsize=11)
ax1.set_ylabel('Rata-rata Selisih Hari (Aktual − Estimasi)', fontsize=11)
ax1.set_title('Rata-rata Keterlambatan\nper Skor Ulasan', fontsize=12, fontweight='bold')
ax1.set_xticks([1, 2, 3, 4, 5])
ax1.set_xticklabels([f'⭐{s}' for s in [1, 2, 3, 4, 5]])
sns.despine(ax=ax1)

ax2 = axes[1]
bars = ax2.bar(delay_score['review_score'], delay_score['pct_terlambat'],
               color=score_colors, edgecolor='white', width=0.6)
for bar, val in zip(bars, delay_score['pct_terlambat']):
    ax2.text(bar.get_x() + bar.get_width() / 2,
             bar.get_height() + 0.5,
             f'{val:.1f}%', ha='center', fontsize=10, fontweight='bold')
ax2.set_xlabel('Skor Ulasan', fontsize=11)
ax2.set_ylabel('% Order Terlambat', fontsize=11)
ax2.set_title('Persentase Order Terlambat\nper Skor Ulasan', fontsize=12, fontweight='bold')
ax2.set_xticks([1, 2, 3, 4, 5])
ax2.set_xticklabels([f'⭐{s}' for s in [1, 2, 3, 4, 5]])
sns.despine(ax=ax2)

plt.suptitle('Analisis Keterlambatan Pengiriman vs Skor Ulasan',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()


[Analisis 2] Keterlambatan Pengiriman vs Skor Ulasan
 review_score  avg_delay  pct_terlambat
          1.0  -4.004198      36.831342
          2.0  -8.627235      18.878955
          3.0 -10.769533       8.764587
          4.0 -12.367107       3.440963
          5.0 -13.381136       1.853257


### 3. Repeat Purchase & Customer Retention
Menganalisis berapa persen pelanggan melakukan lebih dari satu pembelian, serta kategori produk apa yang paling sering memicu pembelian ulang — penting untuk mengevaluasi loyalitas pelanggan.

In [68]:
purchase_count = (
    base.groupby('customer_unique_id')['order_id']
    .nunique().reset_index()
    .rename(columns={'order_id': 'total_orders'})
)

purchase_count['tipe_pelanggan'] = purchase_count['total_orders'].apply(
    lambda x: 'One-time Buyer' if x == 1 else 'Repeat Buyer'
)

retention_rate = (purchase_count['tipe_pelanggan'] == 'Repeat Buyer').mean() * 100

top_repeat_cat = (
    base
    .merge(purchase_count[['customer_unique_id', 'tipe_pelanggan']], on='customer_unique_id', how='left')
    .merge(order_items_df[['order_id', 'product_id']], on='order_id', how='left')
    .merge(products_df[['product_id', 'product_category_name']], on='product_id', how='left')
    .merge(category_translation_df, on='product_category_name', how='left')
    .query("tipe_pelanggan == 'Repeat Buyer'")
    .groupby('product_category_name_english')['order_id']
    .nunique()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
    .rename(columns={'order_id': 'n_orders', 'product_category_name_english': 'category'})
)

print("\n[Analisis 3] Repeat Purchase & Retention")
print(f"  Retention rate : {retention_rate:.1f}%")
print(f"  One-time buyer : {(purchase_count['tipe_pelanggan'] == 'One-time Buyer').sum():,}")
print(f"  Repeat buyer   : {(purchase_count['tipe_pelanggan'] == 'Repeat Buyer').sum():,}")

fig, axes = plt.subplots(1, 2, figsize=(12, 6))
colors = sns.color_palette('muted', 2)

ax1 = axes[0]
type_counts = purchase_count['tipe_pelanggan'].value_counts()
explode = [0.05, 0]
wedges, texts, autotexts = ax1.pie(
    type_counts,
    labels=type_counts.index,
    autopct='%1.1f%%',
    colors=colors,
    explode=explode,
    startangle=140)
for at in autotexts:
    at.set_fontsize(10)
    at.set_fontweight('bold')
ax1.set_title(f'Komposisi Tipe Pelanggan\n(Retention Rate: {retention_rate:.1f}%)',
              fontsize=12, fontweight='bold')

ax2 = axes[1]
bar_colors = sns.color_palette('muted', len(top_repeat_cat))
bars = ax2.barh(top_repeat_cat['category'][::-1],
                top_repeat_cat['n_orders'][::-1],
                color=bar_colors, edgecolor='white')
for bar, val in zip(bars, top_repeat_cat['n_orders'][::-1]):
    ax2.text(bar.get_width() + 5, bar.get_y() + bar.get_height() / 2,
             f'{val:,}', va='center', fontsize=9, fontweight='bold')
ax2.set_xlabel('Jumlah Order', fontsize=11)
ax2.set_title('Top 10 Kategori Produk\nRepeat Buyer', fontsize=12, fontweight='bold')
sns.despine(ax=ax2)

plt.suptitle('Repeat Purchase & Customer Retention',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()


[Analisis 3] Repeat Purchase & Retention
  Retention rate : 3.0%
  One-time buyer : 90,557
  Repeat buyer   : 2,801


### 4. Analisis Geografis Revenue
Memetakan distribusi revenue berdasarkan state/kota asal pelanggan untuk mengidentifikasi pasar yang sudah matang vs pasar dengan potensi pertumbuhan yang belum tergarap.

In [69]:
geo_revenue = (
    base.groupby('customer_state')
    .agg(total_revenue=('order_value', 'sum'),
         n_orders=('order_id', 'count'),
         avg_order_value=('order_value', 'mean'))
    .reset_index()
    .sort_values('total_revenue', ascending=False)
)
geo_revenue['revenue_share_%'] = (
    geo_revenue['total_revenue'] / geo_revenue['total_revenue'].sum() * 100
).round(2)

print("\n[Analisis 4] Revenue per State (Top 10)")
print(geo_revenue.head(10).to_string(index=False))

top10 = geo_revenue.head(10)
top10_avg = geo_revenue.nlargest(10, 'avg_order_value').sort_values('avg_order_value')

fig, axes = plt.subplots(1, 2, figsize=(14, 6))
colors_top = sns.color_palette('muted', 10)

ax1 = axes[0]
bars = ax1.barh(top10['customer_state'][::-1],
                top10['total_revenue'][::-1] / 1_000,
                color=colors_top, edgecolor='white')
for bar, val in zip(bars, top10['total_revenue'][::-1] / 1_000):
    ax1.text(bar.get_width() + 5, bar.get_y() + bar.get_height() / 2,
             f'R${val:,.0f}K', va='center', fontsize=9, fontweight='bold')
ax1.set_xlabel('Total Revenue (Ribuan BRL)', fontsize=11)
ax1.set_title('Top 10 State\nberdasarkan Total Revenue', fontsize=12, fontweight='bold')
ax1.xaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}K'))
sns.despine(ax=ax1)

ax2 = axes[1]
bars = ax2.barh(top10_avg['customer_state'],
                top10_avg['avg_order_value'],
                color=sns.color_palette('muted', len(top10_avg)), edgecolor='white')
for bar, val in zip(bars, top10_avg['avg_order_value']):
    ax2.text(bar.get_width() + 1, bar.get_y() + bar.get_height() / 2,
             f'R${val:,.0f}', va='center', fontsize=9, fontweight='bold')
ax2.set_xlabel('Rata-rata Nilai Order (BRL)', fontsize=11)
ax2.set_title('Top 10 State\nberdasarkan Rata-rata Nilai Order', fontsize=12, fontweight='bold')
ax2.xaxis.set_major_formatter(mticker.FuncFormatter(lambda x, _: f'R${x:,.0f}'))
sns.despine(ax=ax2)

plt.suptitle('Distribusi Geografis Revenue per State',
             fontsize=13, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()


[Analisis 4] Revenue per State (Top 10)
customer_state  total_revenue  n_orders  avg_order_value  revenue_share_%
            SP     5081724.97     40604       125.153309            38.36
            RJ     1762226.87     12380       142.344658            13.30
            MG     1554886.77     11379       136.645291            11.74
            RS      730192.42      5362       136.179116             5.51
            PR      667371.89      4932       135.314657             5.04
            SC      507716.47      3551       142.978448             3.83
            BA      494092.48      3262       151.469185             3.73
            DF      297313.56      2089       142.323389             2.24
            GO      284193.86      1963       144.775273             2.15
            ES      269026.82      2001       134.446187             2.03


**Insight:**
- High Spender hanya ~33% pelanggan, namun berkontribusi paling besar terhadap total revenue — segmen ini paling kritis untuk dipertahankan.
- Mayoritas pelanggan adalah one-time buyer (>90%), menunjukkan tingkat retensi yang sangat rendah dan peluang besar untuk program loyalitas.
- Keterlambatan pengiriman terbukti menjadi pemicu utama ulasan negatif — order dengan skor ⭐1 memiliki rata-rata keterlambatan paling tinggi dibanding skor lainnya.
- State SP (São Paulo) mendominasi revenue secara signifikan, jauh melampaui state lainnya, mencerminkan konsentrasi pasar yang tinggi di satu wilayah.
- Kategori produk yang sering dibeli repeat buyer cenderung merupakan kebutuhan rutin (rumah tangga, kesehatan), bukan produk one-time purchase seperti elektronik besar.

## Conclusion & Recommendation

- **Conclusion pertanyaan 1:** Kategori cama_mesa_banho (bed & bath) dan beleza_saude (health & beauty) secara konsisten berada di posisi teratas revenue 2016–2018, menunjukkan bahwa produk kebutuhan rumah tangga dan perawatan diri adalah tulang punggung bisnis.
- **Conclusion pertanyaan 2:** Revenue mencapai puncak pada November 2017, sangat mungkin dipicu oleh Black Friday. Setelah puncak tersebut terjadi koreksi di awal 2018 sebelum kembali stabil, pola ini menunjukkan ketergantungan pada momen promosi musiman.
- **Conclusion pertanyaan 3:** Terdapat korelasi negatif yang jelas antara waktu pengiriman dan skor ulasan — semakin lama pengiriman (terutama jika melewati estimasi), semakin rendah skor yang diberikan pelanggan.
- **Conclusion pertanyaan 4:** Kartu kredit mendominasi transaksi (74%) sekaligus menghasilkan rata-rata nilai transaksi tertinggi (R$163), menjadikannya metode pembayaran paling strategis untuk dioptimalkan.
- **Conclusion analisis lanjutan 1:** Segmen High Spender perlu diprioritaskan dalam program retensi karena kontribusi revenue-nya tidak proporsional dibanding jumlahnya.
- **Conclusion analisis lanjutan 2:** Keterlambatan — bukan sekadar durasi pengiriman — adalah faktor paling berpengaruh pada kepuasan pelanggan, sehingga akurasi estimasi pengiriman perlu ditingkatkan.
- **Conclusion analisis lanjutan 3:** Tingkat repeat purchase yang rendah mengindikasikan pelanggan belum memiliki alasan kuat untuk kembali, baik dari sisi harga, loyalitas program, maupun pengalaman berbelanja.
- **Conclusion analisis lanjutan 4:** Konsentrasi revenue di state SP dan RJ mengindikasikan potensi ekspansi yang besar di state-state lain yang belum tergarap secara optimal.

**Rekomendasi Action Item:**
- Buat program loyalitas (poin, cashback, member tier) yang secara khusus menarget High Spender dan pelanggan repeat buyer untuk meningkatkan retention rate yang saat ini masih sangat rendah.
- Optimalkan akurasi estimasi pengiriman — jika estimasi tidak bisa dipercepat, prioritaskan transparansi tracking real-time agar pelanggan tidak merasa dikhianati oleh janji pengiriman.
- Kembangkan strategi promosi musiman yang lebih terencana di sekitar November–Desember untuk memaksimalkan lonjakan demand seperti yang terjadi di 2017, sekaligus mengurangi koreksi tajam pasca-puncak.
- Perluas jangkauan ke state di luar SP dan RJ melalui kampanye geo-targeted, terutama di state dengan avg order value tinggi namun volume transaksi masih rendah.
- Tingkatkan penetrasi kategori produk repeat-friendly (kesehatan, rumah tangga) melalui bundling, subscription, atau reminder pembelian ulang untuk mendorong frekuensi order per pelanggan.
- Pertahankan dukungan metode pembayaran Boleto untuk menjaga aksesibilitas bagi segmen unbanked, namun fokuskan insentif (cashback, cicilan 0%) pada pengguna kartu kredit karena nilai transaksinya jauh lebih tinggi.